### PR TITLE
gh-122085: Use include files for C API deprecations

### PR DESCRIPTION
**Pull request by AA-Turner**
----
:books: Documentation preview :books:: https://cpython-previews--109843.org.readthedocs.build/

* Issue: gh-122085

### CODE REVIEW

File: Doc/whatsnew/3.13.rst [hunk 1]
```diff
@@ -1049,6 +1049,95 @@ Deprecated
...
+* Thread Local Storage (TLS) API:
+
+  * :c:func:`PyThread_create_key`: use :c:func:`PyThread_tss_alloc`
+  * :c:func:`PyThread_delete_key`: use :c:func:`PyThread_tss_free`
+  * :c:func:`PyThread_set_key_value`: use :c:func:`PyThread_tss_set`
+  * :c:func:`PyThread_get_key_value`: use :c:func:`PyThread_tss_get`
+  * :c:func:`PyThread_delete_key_value`: use :c:func:`PyThread_tss_delete`
+  * :c:func:`PyThread_ReInitTLS`: unneeded since Python 3.7
+
 Removed
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]: I would prefer to list Pending Removal at the end, after Removal which are done (where it was).

Comment (AA-Turner) [reply Doc/whatsnew/3.13.rst]: Ahh, I just copied the text from 3.12. This style is consistent with the Python API advanced warnings though, and I think makes sense in that these functions haven't been removed just yet, and are so still deprecated. 

If you want to change the position though, we should be consistent and do it for 3.12 too.

Comment (vstinner) [reply Doc/whatsnew/3.13.rst]: IMO things which already happened should be documented before planned changes.

You change the order in 3.13 and leave 3.12 unchanged.

**Comment by vstinner** [comment]
If you move these sub-sections of the C API section, you should also move the other ones of the "main" Python section: moving "Pending Removal in Python 3.14" after Removal. What do you think?

Screenshot 2023-09-26 at 10-18-55 What’s New In Python 3 13

**Comment by AA-Turner** [comment]
Personally, I probably wouldn't as I think semantically it makes more sense to note the pending removals for Python APIs within the *Deprecated* section -- they are a concise list of how quickly readers will need to adapt to the new APIs.

A

**Comment by vstinner** [comment]
> they are a concise list

My concern is more that these lists are getting bigger at every Python version. Sometimes, I'm considering to create a sub-page of What's New in Python 3.13, since the document is so big, that people may give up to just start reading it. Deprecation/Removal only interest a minority of people.

The whole What's New in Python 3.13 takes 18 pages if I print it as PDF with my default browser configuration.

---

I don't have a strong opinon between:

* Deprecations

  * Pending Removals ...

* Removal

VERSUS:

* Deprecations
* Removal
* Pending Removals ...

I would just prefer to have consistency between the (main) Python section and the C API Changes section.

---

If you prefer to have "Pending Removals ..." *under* the Deprecations section, please just make it consistent in the doc.

**Review by blaisep — APPROVED** [review]
The text is cleaner than before, thank you.
I can't confirm that the content is correct, I hope this gets merged soon.

**Comment by hugovk** [comment]
> My concern is more that these lists are getting bigger at every Python version. Sometimes, I'm considering to create a sub-page of What's New in Python 3.13, since the document is so big, that people may give up to just start reading it. Deprecation/Removal only interest a minority of people.

I agree with this.

I want us to have a static page of pending deprecations, so we don't need to worry synchronising across both pages (3.14.rst <-> 3.13.rst <-> 3.12.rst etc.) _and_ branches (main <-> 3.13 <-> 3.12). 

And we sometimes do forget to update them, for example https://github.com/python/cpython/pull/118947. Additionally, at some point a branch goes to security-fix only, and we can no longer update the What's New.

Then, in "What's New in Python 3.x", we should only list the deprecations newly added _in that release_.

---

Compare the pytest page:

* https://docs.pytest.org/en/latest/deprecations.html

As a user of pytest, this is a useful one-stop page to look where I need to update my code.

I copied it for Pillow:

* https://pillow.readthedocs.io/en/stable/deprecations.html

As a maintainer of Pillow, this is a useful one-stop page to look for old deprecations to remove in the next release.

**Comment by blaisep** [comment]
(OK, I can see that there is more to this than I originally thought)

It seems like the current consensus is to have something like
```
Doc/whatsnew/deprecations/removal.rst
Doc/whatsnew/deprecations/removalpending.rst
Doc/whatsnew/deprecations/deprecations.rst
```
and possibly some `Doc/whatsnew/deprecations/index.rst` with a bit of `.. toctree::` sugar sprinkles?
Shall I bring this up in the docs discord?

**Comment by hugovk** [comment]
Follow on from this. We discussed this on Discourse:

* https://discuss.python.org/t/streamline-whats-new-by-moving-deprecations-and-removals-out-of-news/53997

And in the docs-community meeting last month, where we thought about using includes:

* https://docs-community.readthedocs.io/en/latest/monthly-meeting/2024-06.html#python-project-documentation

* https://discuss.python.org/t/streamline-whats-new-by-moving-deprecations-and-removals-out-of-news/53997/8

Please see PR https://github.com/python/cpython/pull/121241 to make a start on this.

**Comment by AA-Turner** [comment]
With apologies for the force-push, I have hijacked my own PR to use the include-file approach for the C API, whilst still doing the synchronisation.

I've split the work into individual commits, which should be somewhat easy to follow.

A

**Comment by hugovk** [comment]
Is it so that we weren't previously duplicating all the C API deprecations across What's New files?

This will mean the What's New pages will become even longer. But I suppose if we do it for the Python deprecations, that's an argument for also doing it for the C API ones too.

**Comment by AA-Turner** [comment]
> Is it so that we weren't previously duplicating all the C API deprecations across What's New files?

Almost all of the deprecations were in both 3.12 and 3.13, it was only 3.14 that was sparse, so I think it is worth both ensuring there's consistency and that we advertise the deprecations in all What's New pages.

a

**Comment by hugovk** [comment]
Would you like to add these includes to the new `deprecations/index.rst` from https://github.com/python/cpython/pull/122352?

---

I also realised we don't want `.. include:: pending-removal-in-3.14.rst` here in `main` (aka 3.14), we only want to list later versions:

https://github.com/python/cpython/blob/b359f66c4c315ca14b2a075ee136145ba6610760/Doc/deprecations/index.rst?plain=1#L4

The backports are correct:

* https://github.com/python/cpython/blob/3.13/Doc/deprecations/index.rst
* https://github.com/python/cpython/blob/3.12/Doc/deprecations/index.rst

Please could you also remove that line at the same time?

**Comment by AA-Turner** [comment]
> Would you like to add these includes to the new `deprecations/index.rst`

Done



> Please could you also remove that line at the same time?

Done

A

**Review by hugovk — APPROVED** [review]
Thanks!

**Pull request merged**